beacon-chain/rpc/beaconv1/state.go

```diff
@@ -2,29 +2,166 @@ package beaconv1
 
 import (
 	"context"
-	"errors"
+	"strconv"
+	"time"
 
 	ptypes "github.com/gogo/protobuf/types"
+	"github.com/pkg/errors"
 	ethpb "github.com/prysmaticlabs/ethereumapis/eth/v1"
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/helpers"
+	"github.com/prysmaticlabs/prysm/beacon-chain/state"
+	"github.com/prysmaticlabs/prysm/shared/bytesutil"
+	"github.com/prysmaticlabs/prysm/shared/params"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
 )
 
 // GetGenesis retrieves details of the chain's genesis which can be used to identify chain.
 func (bs *Server) GetGenesis(ctx context.Context, _ *ptypes.Empty) (*ethpb.GenesisResponse, error) {
-	return nil, errors.New("unimplemented")
+	genesisTime := bs.GenesisTimeFetcher.GenesisTime()
+	var defaultGenesisTime time.Time
+	var gt *ptypes.Timestamp
+	var err error
+	if genesisTime == defaultGenesisTime {
+		gt, err = ptypes.TimestampProto(time.Unix(0, 0))
+	} else {
+		gt, err = ptypes.TimestampProto(genesisTime)
+	}
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not convert genesis time to proto: %v", err)
+	}
+
+	genValRoot := bs.ChainInfoFetcher.GenesisValidatorRoot()
+	return &ethpb.GenesisResponse{
+		GenesisTime:           gt,
+		GenesisValidatorsRoot: genValRoot[:],
+		GenesisForkVersion:    params.BeaconConfig().GenesisForkVersion,
+	}, nil
 }
 
 // GetStateRoot calculates HashTreeRoot for state with given 'stateId'. If stateId is root, same value will be returned.
 func (bs *Server) GetStateRoot(ctx context.Context, req *ethpb.StateRequest) (*ethpb.StateRootResponse, error) {
-	return nil, errors.New("unimplemented")
+	requestedState, err := bs.getState(ctx, req.StateId)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not get state from ID: %v", err)
+	}
+	stateRoot, err := requestedState.HashTreeRoot(ctx)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not get state root: %v", err)
+	}
+	return &ethpb.StateRootResponse{
+		StateRoot: stateRoot[:],
+	}, nil
```

Comment:
This seems expensive to fetch the whole state just to call HTR. 
Have you considered any other options? i.e. looking up the appropriate block and reading the state root value?